rs-matter/src/respond.rs

```diff
@@ -0,0 +1,269 @@
+/*
+ *
+ *    Copyright (c) 2020-2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+use core::fmt::Display;
+use core::pin::pin;
+
+use embassy_futures::select::{select3, select_slice};
+
+use log::{error, info};
+
+use crate::data_model::core::{DataModel, IMBuffer};
+use crate::data_model::objects::DataModelHandler;
+use crate::data_model::subscriptions::Subscriptions;
+use crate::error::{Error, ErrorCode};
+use crate::interaction_model::busy::BusyInteractionModel;
+use crate::secure_channel::busy::BusySecureChannel;
+use crate::secure_channel::core::SecureChannel;
+use crate::transport::exchange::Exchange;
+use crate::utils::buf::BufferAccess;
+use crate::utils::select::Coalesce;
+use crate::Matter;
+
+/// A trait modeling a generic handler for an exchange.
+pub trait ExchangeHandler {
+    async fn handle(&self, exchange: &mut Exchange<'_>) -> Result<(), Error>;
+
+    fn compose<T>(self, other: T) -> CompositeExchangeHandler<Self, T>
+    where
+        T: ExchangeHandler,
+        Self: Sized,
+    {
+        CompositeExchangeHandler(self, other)
+    }
+}
+
+impl<T> ExchangeHandler for &T
+where
+    T: ExchangeHandler,
+{
+    async fn handle(&self, exchange: &mut Exchange<'_>) -> Result<(), Error> {
+        (*self).handle(exchange).await
+    }
+}
+
+/// A struct for composing two exchange handlers into a single one, where each handler is handling one specific protocol (i.e. SC vs IM).
```

Comment:
maybe document that the code explicity:
  - runs handlers in order
  - only runs 2nd handler if first handler returns a `ErrorCode::InvalidProto` error

It does feel somewhat like documenting implementation details, however it also seems the main intent and usage of this.